java-checks/src/main/java/org/sonar/java/checks/SyncGetterAndSetterCheck.java

```diff
@@ -0,0 +1,129 @@
+import com.google.common.collect.Lists;
+import org.sonar.check.Rule;
+import org.sonar.java.model.ModifiersUtils;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.BlockTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Modifier;
+import org.sonar.plugins.java.api.tree.SynchronizedStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import java.util.Collection;
+import java.util.List;
+
+@Rule(key = "S2886")
+public class SyncGetterAndSetterCheck extends IssuableSubscriptionVisitor {
+
+  private static final GetSetPredicate SETTER = new GetSetPredicate() {
+    @Override
+    public String getStartName() {
+      return "set";
+    }
+
+    @Override
+    public boolean apply(MethodTree methodTree) {
+      return methodTree.simpleName().name().startsWith(getStartName()) && methodTree.parameters().size() == 1 && methodTree.returnType().symbolType().is("void");
+    }
+  };
+  private static final GetSetPredicate GETTER = new GetSetPredicate() {
+    @Override
+    public String getStartName() {
+      return "get";
+    }
+
+    @Override
+    public boolean apply(MethodTree methodTree) {
+      return methodTree.simpleName().name().startsWith(getStartName()) && methodTree.parameters().isEmpty() && !methodTree.returnType().symbolType().is("void");
+    }
+  };
+  private static final GetSetPredicate GETTER_BOOLEAN = new GetSetPredicate() {
+    @Override
+    public String getStartName() {
+      return "is";
+    }
+
+    @Override
+    public boolean apply(MethodTree methodTree) {
+      return methodTree.simpleName().name().startsWith(getStartName()) && methodTree.parameters().isEmpty() && !methodTree.returnType().symbolType().is("void");
+    }
+  };
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.METHOD);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    MethodTree methodTree = (MethodTree) tree;
+    checkMethodTree(methodTree, GETTER, SETTER, "set");
+    checkMethodTree(methodTree, GETTER_BOOLEAN, SETTER, "set");
+    checkMethodTree(methodTree, SETTER, GETTER, "get");
+    checkMethodTree(methodTree, SETTER, GETTER_BOOLEAN, "is");
+  }
+
+  private void checkMethodTree(MethodTree methodTree, GetSetPredicate ownPredicate, GetSetPredicate pairPredicate, String startsWith) {
+    if (isSynchronized(methodTree) && ownPredicate.apply(methodTree)) {
+      // Synchronized getter, lookup the setter.
+      Symbol.TypeSymbol owner = (Symbol.TypeSymbol) methodTree.symbol().owner();
+      Collection<Symbol> pairedMethods = owner.lookupSymbols(startsWith + methodTree.symbol().name().substring(ownPredicate.getStartName().length()));
```

Comment:
the string `startsWith` can be removed to only rely on the `pairPredicate.getStartName()`, which does exactly the same.